llvm/include/llvm/Support/ThreadPool.h

```diff
@@ -92,30 +104,32 @@ class ThreadPool {
                      &Group);
   }
 
-  /// Blocking wait for all the threads to complete and the queue to be empty.
-  /// It is an error to try to add new tasks while blocking on this call.
-  /// Calling wait() from a task would deadlock waiting for itself.
-  void wait();
+private:
+  /// Asynchronous submission of a task to the pool. The returned future can be
+  /// used to wait for the task to finish and is *non-blocking* on destruction.
+  template <typename ResTy>
+  std::shared_future<ResTy> asyncImpl(std::function<ResTy()> Task,
+                                      ThreadPoolTaskGroup *Group) {
 
-  /// Blocking wait for only all the threads in the given group to complete.
-  /// It is possible to wait even inside a task, but waiting (directly or
-  /// indirectly) on itself will deadlock. If called from a task running on a
-  /// worker thread, the call may process pending tasks while waiting in order
-  /// not to waste the thread.
-  void wait(ThreadPoolTaskGroup &Group);
+#if LLVM_ENABLE_THREADS
+    /// Wrap the Task in a std::function<void()> that sets the result of the
+    /// corresponding future.
+    auto R = createTaskAndFuture(Task);
 
-  // Returns the maximum number of worker threads in the pool, not the current
-  // number of threads!
-  unsigned getMaxConcurrency() const { return MaxThreadCount; }
+    asyncEnqueue(std::move(R.first), Group);
+    return R.second.share();
 
-  // TODO: misleading legacy name warning!
-  LLVM_DEPRECATED("Use getMaxConcurrency instead", "getMaxConcurrency")
-  unsigned getThreadCount() const { return MaxThreadCount; }
+#else // LLVM_ENABLE_THREADS Disabled
```

Comment:
I tend to concur with @dwblaikie ’s comment regarding !LLVM_ENABLE_THREADS, but we could do this later maybe. Do you happen to know what are the use cases today for LLVM_ENABLE_THREADS=OFF? (in regards to the perf.comment you made)

Reply:
I don't know actually... I think when I was shipping LLVM inside iOS this was compiled with this option, but it's been a while...

Can you clarify which part of @dwblaikie's comment you refer to? Because it went as far as creating a factory function for the default pool implementation, but that would force updating every users to initialize with something like:`std::unique_ptr<ThreadPoolInterface> = llvm::makeThreadPool(options);` (and every following uses to be pointer based).

Reply:
@aganea ping :)

Reply:
@joker-eph I indeed meant the part about splitting `LLVM_ENABLE_THREADS=OFF` into another class/impl.

Reply:
Right so the part about:

> creating a factory function for the default pool implementation, but that would force updating every users to initialize with something like:std::unique_ptr<ThreadPoolInterface> = llvm::makeThreadPool(options); (and every following uses to be pointer based).

makes this non-practical to me right now. Do you have concrete ideas about making this palatable?

Reply:
Yeah, I think that's maybe a nice place to aspire to, but I don't mind if the default implementation does the current macro conditional behaviour.

I feel a bit more strongly about the naming... Seems slightly awkward that the current implementation would keep the unadorned name and the interface would be adorned.

I think I'd be ok with this patch as-is, then a couple of renames shortly after (long enough after that this has baked) - rename the current/implementation to "DefaultThreadPool" or something and give that time to set. Then rename ThreadPoolInterface to ThreadPool?

Reply:
I just split the implementations, but still using a static `using ThreadPool = ` declaration, how does this look?

Reply:
And here: for the followup renaming: https://github.com/llvm/llvm-project/commit/12073a1f3de2552a0b5f48930a755d5119fae9e6

Reply:
I can also squash the renaming here @dwblaikie if you feel it's better to not do this separately by the way

Reply:
I'll send a follow-up PR for the renaming!